First commits

jsf9k opened the pull request:
## 🗣 Description ##

This pull request contains the initial commits for this project.

## 💭 Motivation and context ##

This project's functionality is necessary for the RPT UAT taking place in the COOL.

See also:
- cisagov/guacscanner-docker#2
- cisagov/guacamole-composition#30
- cisagov/ansible-role-guacamole#38
- cisagov/guacamole-packer#53
- cisagov/cool-assessment-terraform#149

## 🧪 Testing ##

I have applied these changes to env0 in our staging COOL environment and verified that they function as expected.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] _All_ future TODOs are captured in issues, which are referenced in code comments.
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [x] Finalize version.

jsf9k committed:
Bare minimum functionality

jsf9k committed:
Allow VNC username, password, and private ssh key to be specified directly

jsf9k committed:
Specify a region when using boto3

jsf9k committed:
Refine docopt/schema validation for VPC ID

jsf9k committed:
Add code to correctly add instances depending on OS type (Windows vs Linux)

jsf9k committed:
Add tests for stale (terminated) instances

jsf9k committed:
Remove two unnecessary lines from test code

jsf9k committed:
Rename two test methods for clarity

jsf9k committed:
Add a couple of assert_not_called() assertions to test code

jsf9k committed:
Add a test for stopped instances

jsf9k committed:
Combine the two terminated instance tests

They are really doing the same thing, and the OS doesn't matter in this case.

jsf9k committed:
Break out removal of connections by ID into its own function

Also add a new function for removing ghost instances, i.e. instances
in the DB that no longer exist in AWS.  Such a thing can happen if
instances are terminated while the Guacamole server is down.

jsf9k committed:
Remove call_count checks

They do not appear to do what I thought they would.

jsf9k committed:
Bump version from 0.0.1 to 0.0.1-rc.1

jsf9k committed:
Fix location of version file

jsf9k committed:
Improve function that checks for ghost instances

jsf9k committed:
Improve handling of input arguments

jsf9k committed:
Skip instances launched from AMIs that match any of a list of regexes

Also pre-compile all regexes where possible.

jsf9k committed:
Bump version from 0.0.1-rc.1 to 0.0.1-rc.2

jsf9k committed:
Fix a bug in the way the IF of the default VPC is retrieved via instance metadata

jsf9k committed:
Bump version from 0.0.1-rc.2 to 0.0.1-rc.3

jsf9k committed:
Add a --oneshot option that runs the update loop only once

This is useful for testing, although it isn't the way the utility is usually run.

jsf9k committed:
Bump version from 0.0.1-rc.3 to 0.0.1-rc.4

jsf9k committed:
Add an input parameter to specify the sleep between loop iterations

jsf9k committed:
Precompile the VPC ID regex

jsf9k committed:
Bump version from 0.0.1-rc.4 to 0.0.1-rc.5

jsf9k committed:
Move DB connection attempt inside the loop

This allows us to wrap it in a try block, so we can repeat the loop if
the connection fails.

jsf9k committed:
Bump version from 0.0.1-rc.5 to 0.0.1-rc.6

jsf9k committed:
Modify the connection name to be something more meaningful

jsf9k committed:
Bump version from 0.0.1-rc.6 to 0.0.1-rc.7

jsf9k committed:
Modify DB connectin string to use key/value pairs instead of a URI

This avoids the need to URI escape the password, for example, in the
event that it contains special characters.

jsf9k committed:
Remove continue statement

This way the sleep will take place before the next DB connection
attempt is made.

jsf9k committed:
Bump version from 0.0.1-rc.7 to 0.0.1-rc.8

jsf9k committed:
Remove unnecessary line

jsf9k committed:
Use a dictionary row factory for the DB connection

Otherwise the results are returned as tuples, which is not what I was
expecting.  It's also inconvenient to have to figure out the index of
the field one wants, so a dictionary is preferred.

jsf9k committed:
Bump version from 0.0.1-rc.8 to 0.0.1-rc.9

jsf9k committed:
Fix handling of return values from fetchone()

fetchone() returns a dictionary, not a list containing a single dictionary.

jsf9k committed:
Bump version from 0.0.1-rc.9 to 0.0.1-rc.10

jsf9k committed:
Fix a typo in a DB query

jsf9k committed:
Bump version from 0.0.1-rc.10 to 0.0.1-rc.11

jsf9k committed:
Fix error in instance ID regex

jsf9k committed:
Use sql.Identifier for a name field

I use it for all the others but somehow didn't use it for this one.

jsf9k committed:
Fix typo in field name

jsf9k committed:
Bump version from 0.0.1-rc.11 to 0.0.1-rc.12

jsf9k committed:
Fix typo in field name

jsf9k committed:
Bump version from 0.0.1-rc.12 to 0.0.1-rc.13

jsf9k committed:
Update README from skeleton content

jsf9k commented:
I still need to add the code to create `guacuser` and give it permission to use all the connections, and I will do that tomorrow.  I went ahead and requested reviews because those items will be more of the same, so reviewers won't be wasting any time by starting their reviews now; in fact, they may find it more comprehensible to do so.

dav3r commented on README.md:
```diff
@@ -6,20 +6,17 @@
-for step-by-step instructions on how to start a new repository from
-a skeleton. This will save you time and effort when configuring a
-new repository!
+This project is a Python utility that continually scans the instances
```
Be a bit more specific here:
```suggestion
This project is a Python utility that continually scans the EC2 instances
```

dav3r commented on setup.py:
```diff
@@ -81,13 +81,20 @@ def get_version(version_file):
     packages=find_packages(where="src"),
     package_dir={"": "src"},
-    package_data={"example": ["data/*.txt"]},
+    # package_data={"guacamole-connection-scanner": ["data/*.txt"]},
```
Why is this commented out?  Can it be removed or set to an empty dict?

jsf9k replied:
I'd like to leave the line in the file, to remind me what to do if I ever add data that needs to be packaged up.  Please see commit 6a2c4e0, where I added an explanatory comment.

dav3r replied:
Got it - that works, thanks!

dav3r commented on src/guacscanner/__init__.py:
```diff
@@ -1,9 +1,10 @@
-from .example import example_div
 
-__all__ = ["example_div"]
+# __all__ = ["example_div"]
```
Can this line be deleted?

jsf9k replied:
Please see commit 890461b.

dav3r replied:
Even better 👍

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,548 @@
+  >0  An error occurred.
+
+Usage:
+  guacscanner [--log-level=LEVEL] [--oneshot] [--sleep=SECONDS] [--postgres-password=PASSWORD|--postgres-password-file=FILENAME] [--private-ssh-key=KEY|--private-ssh-key-file=FILENAME] [--postgres-username=USERNAME|--postgres-username-file=FILENAME] [--rdp-password=PASSWORD|--rdp-password-file=FILENAME] [--rdp-username=USERNAME|--rdp-username-file=FILENAME] [--vnc-password=PASSWORD|--vnc-password ...
```
Shouldn't `postgres-username/file` come immediately after `postgres-password/file`?
```suggestion
  guacscanner [--log-level=LEVEL] [--oneshot] [--sleep=SECONDS] [--postgres-password=PASSWORD|--postgres-password-file=FILENAME] [--postgres-username=USERNAME|--postgres-username-file=FILENAME] [--private-ssh-key=KEY|--private-ssh-key-file=FILENAME] [--rdp-password=PASSWORD|--rdp-password-file=FILENAME] [--rdp-username=USERNAME|--rdp-username-file=FILENAME] [--vnc-password=PASSWORD|--vnc-password-file=FILENAME] [--vnc-username=USERNAME|--vnc-username-file=FILENAME] [--vpc-id=VPC_ID]
```

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,548 @@
+  >0  An error occurred.
+
+Usage:
+  guacscanner [--log-level=LEVEL] [--oneshot] [--sleep=SECONDS] [--postgres-password=PASSWORD|--postgres-password-file=FILENAME] [--private-ssh-key=KEY|--private-ssh-key-file=FILENAME] [--postgres-username=USERNAME|--postgres-username-file=FILENAME] [--rdp-password=PASSWORD|--rdp-password-file=FILENAME] [--rdp-username=USERNAME|--rdp-username-file=FILENAME] [--vnc-password=PASSWORD|--vnc-password ...
```
Shouldn't `postgres-username/file` come immediately after `postgres-password/file`?
```suggestion
  guacscanner [--log-level=LEVEL] [--oneshot] [--sleep=SECONDS] [--postgres-password=PASSWORD|--postgres-password-file=FILENAME] [--postgres-username=USERNAME|--postgres-username-file=FILENAME] [--private-ssh-key=KEY|--private-ssh-key-file=FILENAME] [--rdp-password=PASSWORD|--rdp-password-file=FILENAME] [--rdp-username=USERNAME|--rdp-username-file=FILENAME] [--vnc-password=PASSWORD|--vnc-password-file=FILENAME] [--vnc-username=USERNAME|--vnc-username-file=FILENAME] [--vpc-id=VPC_ID]
```

jsf9k replied:
I believe this is a duplicate suggestion.

dav3r replied:
Oops, my bad. 🤦

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,548 @@
+  --postgres-password-file=FILENAME    The file from which the PostgreSQL password will be read. [default: /run/secrets/postgres-password]
+  --postgres-username=USERNAME    If specified then the specified value will be used when connecting to the PostgreSQL database.  Otherwise, the username will be read from a local file.
+  --postgres-username-file=FILENAME    The file from which the PostgreSQL username will be read. [default: /run/secrets/postgres-username]
+  --private-ssh-key=KEY  If specified then the specified value will be used for the private ssh key.  Otherwise, the ssh key will be read from a local file.
```
Use preferred capitalization:
```suggestion
  --private-ssh-key=KEY  If specified then the specified value will be used for the private SSH key.  Otherwise, the SSH key will be read from a local file.
```

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,548 @@
+  --postgres-username=USERNAME    If specified then the specified value will be used when connecting to the PostgreSQL database.  Otherwise, the username will be read from a local file.
+  --postgres-username-file=FILENAME    The file from which the PostgreSQL username will be read. [default: /run/secrets/postgres-username]
+  --private-ssh-key=KEY  If specified then the specified value will be used for the private ssh key.  Otherwise, the ssh key will be read from a local file.
+  --private-ssh-key-file=FILENAME  The file from which the private ssh key will be read. [default: /run/secrets/private-ssh-key]
```
Use preferred capitalization:
```suggestion
  --private-ssh-key-file=FILENAME  The file from which the private SSH key will be read. [default: /run/secrets/private-ssh-key]
```

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,548 @@
+    "terminated",
+]
+DEFAULT_AMI_SKIP_REGEXES = [
+    re.compile(r"^guacamole-.*$"),
```
Skip Nessus AMIs as well- those instances are only accessible thru a browser on another instance (e.g. Kali or Teamserver):
```suggestion
    re.compile(r"^guacamole-.*$"),
    re.compile(r"^nessus-.*$"),
```

jsf9k committed:
Add code to create guacuser and giver it permission to use connections

jsf9k committed:
Bump version from 0.0.1-rc.13 to 0.0.1-rc.14

jsf9k committed:
Move creation of guacuser inside the connection creation loop

This is to allow for the PostgreSQL database to come up before trying to query it.

jsf9k committed:
Bump version from 0.0.1-rc.14 to 0.0.1-rc.15

jsf9k committed:
Add some code so that the guacuser_id is only looked up once

jsf9k committed:
Bump version from 0.0.1-rc.15 to 0.0.1-rc.16

jsf9k committed:
Modify entity_exists() and get_entity_id() to take the entity type as a parameter

jsf9k committed:
Bump version from 0.0.1-rc.16 to 0.0.1-rc.17

jsf9k committed:
Update tests now that the guacuser code has been added

jsf9k committed:
Add a todo comment to improve the way the Guacamole user is created

jsf9k committed:
Add a TODO comment for improvement of exception handling

dav3r reviewed (changes requested):
This is a partial review, but I wanted to get these to you sooner rather than later.  I'll finish my review a bit later.

jsf9k committed:
Improve capitalization

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Improve comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Correctly alphabetize CLI arguments

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Also ignore instances running Nessus AMIs

Such instances are currently only accessible via a browser on another instance (Kali or Teamserver).

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add a comment explaining why a line is commented out

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
List functions to be exported

These may be useful to others.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add some exception handling for the AMI regex matching

An AttributeError is thrown sometimes, but I'm not sure why.  I
believe it is a temporary problem that only happens while instances
are in flux.  This exception handling code will still log the error,
but keep things moving instead of bombing out the entire process.

jsf9k committed:
Bump version from 0.0.1-rc.17 to 0.0.1-rc.18

dav3r commented on src/guacscanner/__init__.py:
```diff
@@ -0,0 +1,22 @@
+# although this import is not directly used, it populates the value
+# package_name.__version__, which is used to get version information about this
+# Python package.
+# from .guacscanner import example_div
```
This line can be deleted:
```suggestion
```

jsf9k committed:
Remove unnecessary import

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Improve comment about exception thrown by ami.name access

I have been able to determine the exact cause since first encountering
the exception, so it makes sense to update the comment.

jsf9k committed:
Bump version from 0.0.1-rc.18 to 0.0.1

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,758 @@
+            INSERT_USER_QUERY,
+            (
+                entity_id,
+                # guacadmin
+                bytes.fromhex(
+                    "CA458A7D494E3BE824F5E1E175A1556C0F8EEF2C2D7DF3633BEC4A29C4411960"
+                ),
+                bytes.fromhex(
+                    "FE24ADC5E11E2B25288D1704ABE67A79E342ECC26064CE69C5B3177795A82264"
+                ),
```
This section could use some more comments.  Each new user is being created with the same password hash/salt as the default `guacadmin` user, meaning that if someone actually cares about the security of these passwords, they should be changed from the default.  In our case, we don't use passwords since we rely on Kerberos, so couldn't we create a random password?  Maybe `create_random_password` should be an argument to this `add_user()` function?

jsf9k replied:
My main problem with generating random passwords is that I can't for the life of me obtain the password hash above from the password `guacadmin` and the given salt, despite spending hours with [these instructions](https://guacamole.apache.org/doc/gug/jdbc-auth.html#jdbc-auth-schema-users).  I guess that would be ultra secure, since I'd be setting passwords that even I don't know.

If I could generate the example password hash from the known password and salt then I would happily create random passwords.  By Zeus' beard, I beg someone to show me that I am a fool and have misread the instructions linked to in the previous paragraph before I sink further into madness.

Given that we don't use the passwords, are you OK with me adding the above two paragraphs in a comment and creating an issue to create random passwords in the future?

mcdonnnj replied:
```console
>>> pw_str = "guacadmin"
>>> pw_str += "FE24ADC5E11E2B25288D1704ABE67A79E342ECC26064CE69C5B3177795A82264"
>>> hasher = hashlib.sha256()
>>> hasher.update(pw_str.encode())
>>> hasher.hexdigest()
'ca458a7d494e3be824f5e1e175a1556c0f8eef2c2d7df3633bec4a29c4411960'
```

I followed the logic in [the Guacamole source](https://github.com/apache/guacamole-client/blob/62b26fb594aa160478ce44c6d2bb8e6a1cef4174/extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/security/SHA256PasswordEncryptionService.java#L34-L62). It looks like the main thing is they want to concatenate the (uppercase!) string of the hex representation of the salt bytes to the plaintext password and not just concatenate two byte values.

dav3r replied:
I'm fine with you creating an issue for this.  Let me know when you have some time to show me what you've tried here and we can put our noggins together to figure this out.  I spent a bunch of time trying to do this as well back when I was first setting up Guacamole, so I feel your pain.

jsf9k replied:
@mcdonnnj has shown me the light WRT generating the passwords.  I will change the code to generate a random password.

That @mcdonnnj is _smart_!

jsf9k replied:
Please have a look at commit e595bb3.

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,758 @@
+
+    vpc_id = validated_args["--vpc-id"]
+
+    ec2 = boto3.resource("ec2", region_name="us-east-1")
```
Should the region be an input parameter that defaults to `us-east-1`?  Or better yet, have it default to the region that the current instance is running in - something like what we do [here](
https://github.com/cisagov/ansible-role-freeipa-server/blob/ab1a00f2d3cadcd28373f86e05ccfe7aabd40e3c/files/00_setup_freeipa.sh#L154-L162) except we would use http://169.254.169.254/latest/meta-data/placement/region.

jsf9k replied:
Try on commit ec52ab4 for size.

dav3r replied:
Love it!

dav3r commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,498 @@
+@mock_ec2
+def test_addition_of_guacuser():
+    """Verify that adding the guacuser works as expected."""
+    # Create and VPC
```
Typo:
```suggestion
    # Create a VPC
```

dav3r commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,498 @@
+@mock_ec2
+def test_guacuser_already_exists():
+    """Verify that the case where the guacuser already exists works as expected."""
+    # Create and VPC
```
```suggestion
    # Create a VPC
```

dav3r reviewed (approved):
This looks excellent- nice work on the tests too! 🥇 
Please take a look at the comments I left.

jsf9k committed:
Fix typos in comments

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Add a --region CLI optional, defaulting to us-east-1

This option is used in the case where --vpc-id is specified.  It
specifies the AWS region in which the VPC with the given ID exists.
If --vpc-id is not specified then the region where the current
instances exists is read from EC2 metadata.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

felddy commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,498 @@
+        assert return_code == 1, "main() should exit with error"
+
+
+@mock_ec2
```
Whoa..  cool!

felddy reviewed (approved):
Wow!  This is amazing.  The `moto` tests are super cool.  

May the blessed sunlight shine on you like a great peat fire, 
so that stranger and friend may come and warm himself at it.

🌞🔥

jsf9k commented:
> Wow! This is amazing. The `moto` tests are super cool.

It is indeed some cacafuego, even though I haven't (yet) maximized its potential here.

mcdonnnj commented on setup.py:
```diff
@@ -81,13 +81,21 @@ def get_version(version_file):
     package_dir={"": "src"},
-    package_data={"example": ["data/*.txt"]},
+    # This package contains no data
+    # package_data={"guacscanner": ["data/*.txt"]},
```
If there's no package data I would prefer we just remove this and the `src/guacscanner/data/` directory should also be removed.

```suggestion
```

jsf9k replied:
Please see commit c69e775.

mcdonnnj commented on setup.py:
```diff
@@ -81,13 +81,21 @@ def get_version(version_file):
     include_package_data=True,
-    install_requires=["docopt", "schema", "setuptools >= 24.2.0"],
+    install_requires=[
+        "boto3 == 1.19.6",
+        "docopt == 0.6.2",
+        "ec2-metadata == 2.5.0",
+        "psycopg == 3.0.1",
+        "schema == 0.7.4",
```
It would be nice to loosen these up and rely on `pipenv` or similar in the deploying environment (like the Dockerized version) to control exact versions.

jsf9k replied:
Please see the comment I added in commit 478e1a3, which references #9.

mcdonnnj commented on src/guacscanner/__init__.py:
```diff
@@ -0,0 +1,21 @@
+# although this import is not directly used, it populates the value
+# package_name.__version__, which is used to get version information about this
+# Python package.
+from . import guacscanner  # noqa: F401
```
Would you update the comment to include this import? It currently only discusses the import for `_version`.

jsf9k replied:
Please see commit cd170b4.

mcdonnnj commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,764 @@
+    instance,
+    add_instance_states,
+    remove_instance_states,
+    vnc_username,
+    vnc_password,
+    private_ssh_key,
+    rdp_username,
+    rdp_password,
```
Not super critical or anything, but as a thought would this make sense as a [NamedTuple](https://docs.python.org/3/library/typing.html#typing.NamedTuple) to pass around? That would make any future changes in data for connections much less ugly to implement.

jsf9k replied:
I actually prefer it to be explicit.  To my mind, using a named tuple obfuscates things a bit.  What say you @dav3r and @felddy?

dav3r replied:
You both make fair points- I can't think of a good tie-breaker reason, so I'm cool either way.

mcdonnnj replied:
I was just thinking of what I did in [cisagov/hash-http-content](https://github.com/cisagov/hash-http-content/blob/25057b67ad54b083a6f1b3e89d67482a78fcf960/src/hash_http_content/hasher.py#L71-L79) as I felt like it made it easier to set expectations. If we needed to change information for connections at a later point I feel like it's far more maintainable to modify a NamedTuple than update every function signature and call to reflect changes.

jsf9k replied:
Please see commit 95de413.

mcdonnnj commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,764 @@
+                    error="Possible values for --vpc-id are the characters vpc- followed by either 8 or 17 hexadecimal digits.",
+                ),
+            ),
+            str: object,  # Don't care about other keys, if any
```
Not necessarily required for this iteration, but since you added a `--region` option it would be nice to check that it's a valid region.

jsf9k replied:
I added a TODO comment in commit a6278b8 that references #6.

mcdonnnj commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,764 @@
+    oneshot = validated_args["--oneshot"]
+    logging.debug("oneshot is %s.", oneshot)
+
+    postgres_password = validated_args["--postgres-password"]
+    if postgres_password is None:
+        with open(validated_args["--postgres-password-file"], "r") as file:
+            postgres_password = file.read()
+
+    postgres_username = validated_args["--postgres-username"]
+    if postgres_username is None:
+        with open(validated_args["--postgres-username-file"], "r") as file:
+            postgres_username = file.read()
+
+    rdp_password = validated_args["--rdp-password"]
+    if rdp_password is None:
+        with open(validated_args["--rdp-password-file"], "r") as file:
+            rdp_password = file.read()
+
+    rdp_username = validated_args["--rdp-username"]
+    if rdp_username is None:
+        with open(validated_args["--rdp-username-file"], "r") as file:
+            rdp_username = file.read()
+
+    vnc_password = validated_args["--vnc-password"]
+    if vnc_password is None:
+        with open(validated_args["--vnc-password-file"], "r") as file:
+            vnc_password = file.read()
+
+    vnc_username = validated_args["--vnc-username"]
+    if vnc_username is None:
+        with open(validated_args["--vnc-username-file"], "r") as file:
+            vnc_username = file.read()
+
+    private_ssh_key = validated_args["--private-ssh-key"]
+    if private_ssh_key is None:
+        with open(validated_args["--private-ssh-key-file"], "r") as file:
+            private_ssh_key = file.read()
```
The `docopt` specification has these all marked as optional, but this logic implies that at least one of each pair is required.

Please see the following example:

```console
$ guacscanner
Traceback (most recent call last):
  File "/home/mcdonnnj/.pyenv/versions/guacscanner/bin/guacscanner", line 33, in <module>
    sys.exit(load_entry_point('guacscanner', 'console_scripts', 'guacscanner')())
  File "/home/mcdonnnj/code/guacscanner/src/guacscanner/guacscanner.py", line 625, in main
    with open(validated_args["--postgres-password-file"], "r") as file:
FileNotFoundError: [Errno 2] No such file or directory: '/run/secrets/postgres-password'
```

If I change that pair to `()` instead of `[]` I am correctly given the usage output instead.

jsf9k replied:
You shouldn't get the usage output in this case.  If you don't define either parameter explicitly then you get the default value for the `--postgres-password-file` option, which is `/run/secrets/postgres-password`.  If that file does not exist, then you get the error you showed in your comment.

I don't _think_ any changes are necessary here, but I'm open to being convinced otherwise.

mcdonnnj replied:
Ah, you're right. I missed the defaults in the giant wall of text. I suppose #5 would resolve the not-very-helpful output if you neglect to give an argument and it bombs out like this.

jsf9k replied:
Yes

mcdonnnj commented:
I have a review in progress with at least one group of important (to me!) changes. I just wanted to put a pin here since you have two approvals now.

jsf9k committed:
Add code to correctly handle generation of Guacamole passwords

I also added type hints to the add_user() function, since the types of
the password and salt input variables are important to note.

Thanks to @mcdonnnj for figuring out exactly how the salted password
hash should be generated!

mcdonnnj commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,498 @@
+def test_new_linux_instance():
+    """Verify that adding a new Linux instance works as expected."""
+    # Create and populate a VPC with an EC2 instance
+    ec2 = boto3.client("ec2", "us-east-1")
+    vpc = ec2.create_vpc(CidrBlock="10.19.74.0/24")
+    vpc_id = vpc["Vpc"]["VpcId"]
+    subnet = ec2.create_subnet(CidrBlock="10.19.74.0/24", VpcId=vpc_id)
+    subnet_id = subnet["Subnet"]["SubnetId"]
```
Just as a thought: this logic is used in a number of tests so it would it be useful to break it out into a helper function?

jsf9k replied:
I added a TODO comment in commit 5c45b4a that references #7.

mcdonnnj commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,498 @@
+        name="Mock PostgreSQL connection", spec_set=psycopg.Connection
+    )
+    mock_cursor = MagicMock(name="Mock PostgreSQL cursor", spec_set=psycopg.Cursor)
+    mock_cursor.__enter__.return_value = mock_cursor
+    mock_connection.__enter__.return_value = mock_connection
+    mock_connection.cursor.return_value = mock_cursor
```
Are there no `side_effect`s to use in this test?

jsf9k replied:
No, none were necessary here.

mcdonnnj commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,764 @@
+]
+
+# Some precompiled regexes
+INSTANCE_ID_REGEX = re.compile(r"^.* \((?P<id>i-[0-9a-f]{17})\)$")
```
This might benefit from a succinct comment just to mention the `(?P<value>)` usage.

jsf9k replied:
Please enjoy commit c5f57d1.

mcdonnnj reviewed (changes requested):
Great work hammering this out. I have some feedback, not all of which requires immediate action for approval. I do think it would good to put in an issue to break up the project a bit. Everything in `guacscanner.py` is a lot going on.

jsf9k committed:
Uppercase values that are saved to the DB before converting to byte arrays

This is necessary to match the output of the Guacamole DB
initialization script.

jsf9k committed:
Get rid of need for noqa in one line

jsf9k committed:
Remove package_data code that is commented out anyway

Also remove the src/guacscanner/data directory.

jsf9k committed:
Add todo comment re validating AWS region names passed in on the command line

jsf9k committed:
Add a comment explaining the use of a named capture group

jsf9k committed:
Add a TODO comment to reduce duplication of mock EC2 setup code

This comment references #7.

jsf9k committed:
Add TODO comment for loosening version pins

jsf9k committed:
Remove debug logging statement that could contain sensitive information

jsf9k committed:
Avoid importing from psycopg in two different ways

LGTM complains about this on the grounds that it is confusing.

jsf9k committed:
Add an LGTM comment string to disable a warning

We must use the same password hashing algorithm as is used in the
Guacamole source code, so we cannot avoid the LGTM warning.

jsf9k committed:
Do not include username in logging statement

Our linters think this is logging of sensitive information, which is
reasonable.

jsf9k committed:
Bump version from 0.0.1 to 0.0.2

jsf9k committed:
Bump version from 0.0.2 to 0.0.2-rc.1

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,811 @@
+    "running",
+]
+DEFAULT_PASSWORD_LENGTH = 32
+DEFAULT_SALT_LENGTH = 32
```
Preserve alphabetical order:
```suggestion
DEFAULT_PASSWORD_SALT_LENGTH = 32
```
Or re-order the lines if you prefer, but doing it this way keeps the two password-related lines next to each other.  Note that you will also need to change where this variable is used below.

dav3r commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,811 @@
+        )
+    if salt is None:
+        # Generate a random byte array
+        salt = secrets.token_bytes(DEFAULT_SALT_LENGTH)
```
Related to suggestion above:
```suggestion
        salt = secrets.token_bytes(DEFAULT_PASSWORD_SALT_LENGTH)
```

dav3r commented on tests/test_guacscanner.py:
```diff
@@ -0,0 +1,501 @@
+    """Verify that adding a new Linux instance works as expected."""
+    # Create and populate a VPC with an EC2 instance
+    #
+    # TODO: Create a test fixture to reduce duplicatin of this EC2
```
Typo:
```suggestion
    # TODO: Create a test fixture to reduce duplication of this EC2
```

dav3r reviewed (approved):
👍 👍

jsf9k committed:
Rename a variable to preserve alphabetical order and yet keep related variables together

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Fix a typo

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Use a dataclass for passing Guacamole connection parameters as function inputs

This simplifies function signatures and makes modifying the list of
connection parameters less error-prone.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on src/guacscanner/guacscanner.py:
```diff
@@ -0,0 +1,805 @@
+        try:
+            with psycopg.connect(
+                db_connection_string, row_factory=psycopg.rows.dict_row
+            ) as db_connection:
```
The `connection` object is not a true context manager and the connection is not closed when you leave the context. Should the database connection be made before entering the loop so it is re-used and then closed upon exiting the loop? The example in the [psycopg documentation](https://www.psycopg.org/docs/connection.html#connection) highlights this fact:

```python
conn = psycopg2.connect(DSN)

with conn:
    with conn.cursor() as curs:
        curs.execute(SQL1)

with conn:
    with conn.cursor() as curs:
        curs.execute(SQL2)

# leaving contexts doesn't close the connection
conn.close()
```

jsf9k replied:
Please have a look at commit 270ac59.

mcdonnnj replied:
Did you mean to move the `sleep` call to the top of the loop?

https://github.com/cisagov/guacscanner/blob/51bf9d2507be348deef25e561aac563fb6658183/src/guacscanner/guacscanner.py#L720-L721

jsf9k replied:
Yes.  This way the sleep will still happen if an exception is thrown when trying to open a connection to the DB.  The only other way I saw of making that happen was to call `time.sleep()` from both the exception handler and the bottom of the loop and this seemed cleaner.

The sleep needs to happen in this case since otherwise the loop will relentlessly and continuously try to reconnect to the DB.

mcdonnnj reviewed (approved):
Strong work again 💪💪💪Thank you for taking my suggestions into consideration. I had one question below that isn't an immediate blocker.

jsf9k committed:
Do not create DB connection in a context manager

@mcdonnnj pointed out that psycopg.connect() does not entirely act as
expected in that it does not automatically close the connection when
you exit the context.  It only rolls back transactions in the event
that one fails.

Since we are managing transactions in this code via explicit calls to
psycopg.Connection.commit(), the context manager isn't actually buying
us anything; therefore, I have gotten rid of it.

I added an explicit psycopg.Connection.close() at the bottom of the
loop.  I don't want to create the connection outside the loop and
leave it open because the guacscanner may be up and running for months
at a time, and who knows what could go wrong.  It seems safer to just
recreate the connection at the top of the loop.

jsf9k committed:
Bump version from 0.0.2-rc.1 to 0.0.2-rc.2

jsf9k committed:
Bump version from 0.0.2-rc.2 to 0.0.2

jsf9k committed:
Bump version from 0.0.2 to 1.0.0
